### ISSUE TITLE
High memory usage when loading a custom-strided/non-contigous numpy array 

### ISSUE THREAD

**Issue by FirefoxMetzger** [dask/dask#7208]
**What happened**: 

I am trying to load an array from numpy into dask that uses custom striding (essentially a variant of [stride_tricks](https://numpy.org/doc/stable/reference/generated/numpy.lib.stride_tricks.as_strided.html)). The problem is that calling `dask.array.from_array` creates an inflated copy of the array which blows up memory. I admit that I have not used dask before, so it can very well be user-error on my part.

The background is that I am trying to create a (naive) implementation of convolution for ndimages. In my case, both the image (stack of HD resolution images) and the kernel (kernel size >50x50) are large. Inflating the image (as is done in some convolution implementations to then rely on BLAS for matrix-matrix or matrix-vector math), is not admissible. However, we can express a windowed version of the array/image using custom strides without creating a copy/inflating. After having done that, we can apply the kernel without copying data around. After vectorization (SIMD) along the last dimension (it remains contiguous), doing so is surprisingly fast. I have a working implementation that makes it concurrent using OpenMP (apply kernel to windows in parallel), but I was musing about creating a dask-based implementation, which could be much better/cleaner.

**What you expected to happen**:

The array gets loaded into dask without creating a copy and while preserving the original strides.

**Minimal Complete Verifiable Example**:

```python
import numpy as np
from skimage.util import view_as_windows  # cleaner than using stride_tricks directly
import dask.array as da

size= 10000

# okay, only stores 10000x10000 doubles
numpy_array = view_as_windows(np.arange(size*size).reshape(size,size), (5,5))

# not okay, store 2498000400 doubles (~30x more memory needed)
foo = da.from_array(view_as_windows(np.arange(size*size).reshape(size,size), (5,5)))
```

**Anything else we need to know?**:

**Environment**:

- Dask version: 2.21.0
- Python version: 3.7.0 64-bit
- Operating System: Windows 10 V20H2
- Install method (conda, pip, source): pip

**Comment by jsignell** [comment]
hmm I am having trouble reproducing this. When I run your examples I see the same size and shape. Where you looking at something else?

```python
In [1]: import numpy as np
   ...: from skimage.util import view_as_windows  # cleaner than using stride_tricks directly
   ...: import dask.array as da
   ...: 
   ...: size= 10000
   ...: 
   ...: numpy_array = view_as_windows(np.arange(size*size).reshape(size,size), (5,5))
   ...: 
   ...: foo = da.from_array(numpy_array)

In [2]: numpy_array.shape, numpy_array.size
Out[2]: ((9996, 9996, 5, 5), 2498000400)

In [3]: foo.shape, foo.size
Out[3]: ((9996, 9996, 5, 5), 2498000400)
```

**Comment by FirefoxMetzger** [comment]
Yes, the size and shape are exactly the same for both arrays. The amount of memory consumed, however, is not; that's the cool thing about custom striding. I don't know if I can show this when directly comparing to a dask array (couldn't find a strides attribute), so here is the comparison in pure numpy

```python
import numpy as np
from skimage.util import view_as_windows
size = 5000  # my workstation can't handle size=10000 inflated :(

# occupies ~ 100Mb in memory
strided_array = view_as_windows(np.arange(size*size).reshape(size,size), (5,5))

# occupies ~ 2.5Gb in memory
normal_array = strided_array.copy() 

print(strided_array.strides)  # (400, 4, 400, 4)
print(normal_array.strides) # (9600, 100, 20, 4)
```

The difference is subtle but really important. If you monitor the actual memory used by the process (task-manager / `htop` / etc.), you should see that creating `numpy_array` barely allocates additional memory, but that creating `foo` or `normal_array` significantly increases the memory consumed.

**Comment by dcherian** [comment]
:) #7234 implements the `as_strided` trick with `map_overlap`. Maybe you can test it out.

To be clear, are you concerned about memory usage while computing or memory usage when calling `from_array`. If the latter, I don't see it when profiling 
```
%load_ext memory_profiler
%mprun -f da.from_array da.from_array(numpy_array)
```

<details>
<summary>Details</summary>

```
Filename: /home/deepak/miniconda3/envs/dcpy/lib/python3.8/site-packages/dask/array/core.py

Line #    Mem usage    Increment  Occurences   Line Contents
============================================================
  2899    232.6 MiB    232.6 MiB           1   def from_array(
  2900                                             x,
  2901                                             chunks="auto",
  2902                                             name=None,
  2903                                             lock=False,
  2904                                             asarray=None,
  2905                                             fancy=True,
  2906                                             getitem=None,
  2907                                             meta=None,
  2908                                             inline_array=False,
  2909                                         ):
...
  3036    232.6 MiB      0.0 MiB           1       if isinstance(x, Array):
  3037                                                 raise ValueError(
  3038                                                     "Array is already a dask array. Use 'asarray' or " "'rechunk' instead."
  3039                                                 )
  3040    232.6 MiB      0.0 MiB           1       elif is_dask_collection(x):
  3041                                                 warnings.warn(
  3042                                                     "Passing an object to dask.array.from_array which is already a "
  3043                                                     "Dask collection. This can lead to unexpected behavior."
  3044                                                 )
  3045                                         
  3046    232.6 MiB      0.0 MiB           1       if isinstance(x, (list, tuple, memoryview) + np.ScalarType):
  3047                                                 x = np.array(x)
  3048                                         
  3049    232.6 MiB      0.0 MiB           1       if asarray is None:
  3050    232.6 MiB      0.0 MiB           1           asarray = not hasattr(x, "__array_function__")
  3051                                         
  3052    232.6 MiB      0.0 MiB           1       previous_chunks = getattr(x, "chunks", None)
  3053                                         
  3054    232.6 MiB      0.0 MiB           2       chunks = normalize_chunks(
  3055    232.6 MiB      0.0 MiB           1           chunks, x.shape, dtype=x.dtype, previous_chunks=previous_chunks
  3056                                             )
  3057                                         
  3058    232.6 MiB      0.0 MiB           1       if name in (None, True):
  3059    232.6 MiB      0.0 MiB           1           token = tokenize(x, chunks)
  3060    232.6 MiB      0.0 MiB           1           original_name = "array-original-" + token
  3061    232.6 MiB      0.0 MiB           1           name = name or "array-" + token
  3062                                             elif name is False:
  3063                                                 original_name = name = "array-" + str(uuid.uuid1())
  3064                                             else:
  3065                                                 original_name = name
  3066                                         
  3067    232.6 MiB      0.0 MiB           1       if lock is True:
  3068                                                 lock = SerializableLock()
  3069                                         
  3070    232.6 MiB      0.0 MiB           1       is_ndarray = type(x) is np.ndarray
  3071    232.6 MiB      0.0 MiB           4       is_single_block = all(len(c) == 1 for c in chunks)
  3072                                             # Always use the getter for h5py etc. Not using isinstance(x, np.ndarray)
  3073                                             # because np.matrix is a subclass of np.ndarray.
  3074    232.6 MiB      0.0 MiB           1       if is_ndarray and not is_single_block and not lock:
  3075                                                 # eagerly slice numpy arrays to prevent memory blowup
  3076                                                 # GH5367, GH5601
  3077    232.6 MiB      0.0 MiB           1           slices = slices_from_chunks(chunks)
  3078    232.6 MiB      0.0 MiB          11           keys = product([name], *(range(len(bds)) for bds in chunks))
  3079    232.6 MiB      0.0 MiB           7           values = [x[slc] for slc in slices]
  3080    232.6 MiB      0.0 MiB           1           dsk = dict(zip(keys, values))
  3081                                         
  3082                                             elif is_ndarray and is_single_block:
  3083                                                 # No slicing needed
  3084                                                 dsk = {(name,) + (0,) * x.ndim: x}
  3085                                             else:
  3086                                                 if getitem is None:
  3087                                                     if fancy:
  3088                                                         getitem = getter
  3089                                                     else:
  3090                                                         getitem = getter_nofancy
  3091                                         
  3092                                                 if inline_array:
  3093                                                     get_from = x
  3094                                                 else:
  3095                                                     get_from = original_name
  3096                                         
  3097                                                 dsk = getem(
  3098                                                     get_from,
  3099                                                     chunks,
  3100                                                     getitem=getitem,
  3101                                                     shape=x.shape,
  3102                                                     out_name=name,
  3103                                                     lock=lock,
  3104                                                     asarray=asarray,
  3105                                                     dtype=x.dtype,
  3106                                                 )
  3107                                                 if not inline_array:
  3108                                                     dsk[original_name] = x
  3109                                         
  3110                                             # Workaround for TileDB, its indexing is 1-based,
  3111                                             # and doesn't seems to support 0-length slicing
  3112    232.6 MiB      0.0 MiB           1       if x.__class__.__module__.split(".")[0] == "tiledb" and hasattr(x, "_ctx_"):
  3113                                                 return Array(dsk, name, chunks, dtype=x.dtype)
  3114                                         
  3115    232.6 MiB      0.0 MiB           1       if meta is None:
  3116    232.6 MiB      0.0 MiB           1           meta = x
  3117                                         
  3118    232.6 MiB      0.0 MiB           1       return Array(dsk, name, chunks, meta=meta, dtype=getattr(x, "dtype", None))


```


</details>

**Comment by FirefoxMetzger** [comment]
Yes, @dcherian I am talking about the `from_array` call. I am getting the same results as you when I try the snippet you posted:

<details>
<summary>mprun Output</summary>

Line #    Mem usage    Increment  Occurences   Line Contents
============================================================
  2683    321.0 MiB    321.0 MiB           1   def from_array(
  2684                                             x,
  2685                                             chunks="auto",
  2686                                             name=None,
  2687                                             lock=False,
  2688                                             asarray=None,
  2689                                             fancy=True,
  2690                                             getitem=None,
  2691                                             meta=None,
  2692                                         ):
  2693                                             """ Create dask array from something that looks like an array
  2694                                         
  2695                                             Input must have a ``.shape``, ``.ndim``, ``.dtype`` and support numpy-style slicing.
  2696                                         
  2697                                             Parameters
  2698                                             ----------
  2699                                             x : array_like
  2700                                             chunks : int, tuple
  2701                                                 How to chunk the array. Must be one of the following forms:
  2702                                         
  2703                                                 - A blocksize like 1000.
  2704                                                 - A blockshape like (1000, 1000).
  2705                                                 - Explicit sizes of all blocks along all dimensions like
  2706                                                   ((1000, 1000, 500), (400, 400)).
  2707                                                 - A size in bytes, like "100 MiB" which will choose a uniform
  2708                                                   block-like shape
  2709                                                 - The word "auto" which acts like the above, but uses a configuration
  2710                                                   value ``array.chunk-size`` for the chunk size
  2711                                         
  2712                                                 -1 or None as a blocksize indicate the size of the corresponding
  2713                                                 dimension.
  2714                                             name : str, optional
  2715                                                 The key name to use for the array. Defaults to a hash of ``x``.
  2716                                                 By default, hash uses python's standard sha1. This behaviour can be
  2717                                                 changed by installing cityhash, xxhash or murmurhash. If installed,
  2718                                                 a large-factor speedup can be obtained in the tokenisation step.
  2719                                                 Use ``name=False`` to generate a random name instead of hashing (fast)
  2720                                         
  2721                                                 .. note::
  2722                                         
  2723                                                    Because this ``name`` is used as the key in task graphs, you should
  2724                                                    ensure that it uniquely identifies the data contained within. If
  2725                                                    you'd like to provide a descriptive name that is still unique, combine
  2726                                                    the descriptive name with :func:`dask.base.tokenize` of the
  2727                                                    ``array_like``. See :ref:`graphs` for more.
  2728                                         
  2729                                             lock : bool or Lock, optional
  2730                                                 If ``x`` doesn't support concurrent reads then provide a lock here, or
  2731                                                 pass in True to have dask.array create one for you.
  2732                                             asarray : bool, optional
  2733                                                 If True then call np.asarray on chunks to convert them to numpy arrays.
  2734                                                 If False then chunks are passed through unchanged.
  2735                                                 If None (default) then we use True if the ``__array_function__`` method
  2736                                                 is undefined.
  2737                                             fancy : bool, optional
  2738                                                 If ``x`` doesn't support fancy indexing (e.g. indexing with lists or
  2739                                                 arrays) then set to False. Default is True.
  2740                                             meta : Array-like, optional
  2741                                                 The metadata for the resulting dask array.  This is the kind of array
  2742                                                 that will result from slicing the input array.
  2743                                                 Defaults to the input array.
  2744                                         
  2745                                             Examples
  2746                                             --------
  2747                                         
  2748                                             >>> x = h5py.File('...')['/data/path']  # doctest: +SKIP
  2749                                             >>> a = da.from_array(x, chunks=(1000, 1000))  # doctest: +SKIP
  2750                                         
  2751                                             If your underlying datastore does not support concurrent reads then include
  2752                                             the ``lock=True`` keyword argument or ``lock=mylock`` if you want multiple
  2753                                             arrays to coordinate around the same lock.
  2754                                         
  2755                                             >>> a = da.from_array(x, chunks=(1000, 1000), lock=True)  # doctest: +SKIP
  2756                                         
  2757                                             If your underlying datastore has a ``.chunks`` attribute (as h5py and zarr
  2758                                             datasets do) then a multiple of that chunk shape will be used if you
  2759                                             do not provide a chunk shape.
  2760                                         
  2761                                             >>> a = da.from_array(x, chunks='auto')  # doctest: +SKIP
  2762                                             >>> a = da.from_array(x, chunks='100 MiB')  # doctest: +SKIP
  2763                                             >>> a = da.from_array(x)  # doctest: +SKIP
  2764                                         
  2765                                             If providing a name, ensure that it is unique
  2766                                         
  2767                                             >>> import dask.base
  2768                                             >>> token = dask.base.tokenize(x)  # doctest: +SKIP
  2769                                             >>> a = da.from_array('myarray-' + token)  # doctest: +SKIP
  2770                                             """
  2771    321.0 MiB      0.0 MiB           1       if isinstance(x, Array):
  2772                                                 raise ValueError(
  2773                                                     "Array is already a dask array. Use 'asarray' or " "'rechunk' instead."
  2774                                                 )
  2775    321.0 MiB      0.0 MiB           1       elif is_dask_collection(x):
  2776                                                 warnings.warn(
  2777                                                     "Passing an object to dask.array.from_array which is already a "
  2778                                                     "Dask collection. This can lead to unexpected behavior."
  2779                                                 )
  2780                                         
  2781    321.0 MiB      0.0 MiB           1       if isinstance(x, (list, tuple, memoryview) + np.ScalarType):
  2782                                                 x = np.array(x)
  2783                                         
  2784    321.0 MiB      0.0 MiB           1       if asarray is None:
  2785    321.0 MiB      0.0 MiB           1           asarray = not hasattr(x, "__array_function__")
  2786                                         
  2787    321.0 MiB      0.0 MiB           1       previous_chunks = getattr(x, "chunks", None)
  2788                                         
  2789    321.0 MiB      0.0 MiB           1       chunks = normalize_chunks(
  2790    321.1 MiB      0.0 MiB           1           chunks, x.shape, dtype=x.dtype, previous_chunks=previous_chunks
  2791                                             )
  2792                                         
  2793    321.1 MiB      0.1 MiB           1       if name in (None, True):
  2794    321.1 MiB      0.0 MiB           1           token = tokenize(x, chunks)
  2795    321.1 MiB      0.0 MiB           1           original_name = "array-original-" + token
  2796    321.1 MiB      0.0 MiB           1           name = name or "array-" + token
  2797                                             elif name is False:
  2798                                                 original_name = name = "array-" + str(uuid.uuid1())
  2799                                             else:
  2800                                                 original_name = name
  2801                                         
  2802    321.1 MiB      0.0 MiB           1       if lock is True:
  2803                                                 lock = SerializableLock()
  2804                                         
  2805                                             # Always use the getter for h5py etc. Not using isinstance(x, np.ndarray)
  2806                                             # because np.matrix is a subclass of np.ndarray.
  2807    321.1 MiB      0.0 MiB           4       if type(x) is np.ndarray and all(len(c) == 1 for c in chunks):
  2808                                                 # No slicing needed
  2809                                                 dsk = {(name,) + (0,) * x.ndim: x}
  2810                                             else:
  2811    321.1 MiB      0.0 MiB           1           if getitem is None:
  2812    321.1 MiB      0.0 MiB           1               if type(x) is np.ndarray and not lock:
  2813                                                         # simpler and cleaner, but missing all the nuances of getter
  2814    321.1 MiB      0.0 MiB           1                   getitem = operator.getitem
  2815                                                     elif fancy:
  2816                                                         getitem = getter
  2817                                                     else:
  2818                                                         getitem = getter_nofancy
  2819                                         
  2820    321.1 MiB      0.0 MiB           1           dsk = getem(
  2821    321.1 MiB      0.0 MiB           1               original_name,
  2822    321.1 MiB      0.0 MiB           1               chunks,
  2823    321.1 MiB      0.0 MiB           1               getitem=getitem,
  2824    321.1 MiB      0.0 MiB           1               shape=x.shape,
  2825    321.1 MiB      0.0 MiB           1               out_name=name,
  2826    321.1 MiB      0.0 MiB           1               lock=lock,
  2827    321.1 MiB      0.0 MiB           1               asarray=asarray,
  2828    321.1 MiB      0.0 MiB           1               dtype=x.dtype,
  2829                                                 )
  2830    321.1 MiB      0.0 MiB           1           dsk[original_name] = x
  2831                                         
  2832                                             # Workaround for TileDB, its indexing is 1-based,
  2833                                             # and doesn't seems to support 0-length slicing
  2834    321.1 MiB      0.0 MiB           1       if x.__class__.__module__.split(".")[0] == "tiledb" and hasattr(x, "_ctx_"):
  2835                                                 return Array(dsk, name, chunks, dtype=x.dtype)
  2836                                         
  2837    321.1 MiB      0.0 MiB           1       if meta is None:
  2838    321.1 MiB      0.0 MiB           1           meta = x
  2839                                         
  2840    321.1 MiB      0.0 MiB           1       return Array(dsk, name, chunks, meta=meta, dtype=getattr(x, "dtype", None))
</details>

However, when I monitor the system's memory usage in the task manager I see a (temporary) spike of several GB when calling `from_array`. This also happens during the call that is being profiled by `mprun`.

![dask_array_running](https://user-images.githubusercontent.com/4402489/108755019-001d9a80-7547-11eb-929c-56a031ef6d4f.png)

For comparison, my base usage (after having reset the kernel and restarted VSCode to free any residual memory) is around 8GB, so the call to `from_array` temporarily consumes ~6 GB for `size=8000`, which is very close to the size of the default-strided (I call it "inflated") array of 6.1GB.
![base_memory_usage](https://user-images.githubusercontent.com/4402489/108755123-23e0e080-7547-11eb-95ce-44accaa73cc6.png)

Apparently, the spike is temporary and vanishes after the array has been created

![dask_spike](https://user-images.githubusercontent.com/4402489/108755359-715d4d80-7547-11eb-9ec7-13df74fc9b33.png)

I'm trying to come up with a good way to measure it in python, but so far I have nothing. Ideas are very welcome.

**Comment by jakirkham** [comment]
Sorry if I missed it, but did you try the linked PR as well?

**Comment by FirefoxMetzger** [comment]
@jakirkham No, I didn't get to play around with the PR yet. I hope to give it a shot tomorrow and will report back if it improves the situation.

**Comment by FirefoxMetzger** [comment]
Using #7234 I see the same spike when using `da.from_array` on a custom strided array. However, doing something like

```
from dask.array.overlap import sliding_window_view
dask_overlap = sliding_window_view(da.from_array(np.arange(size*size).reshape(size,size)), (5,5), (0,1))
```

works without producing a temporary memory spike.

**Comment by dcherian** [comment]
Yes `dask_overlap` is the intended use. Can you apply your kernel and confirm that the memory use is as expected when you compute? Memory usage for `func(dask_overlap)` should be much less than memory usage of `func(dask_overlap.map_blocks(np.copy))`

**Comment by bmerry** [comment]
It seems to be caused by the hashing used to generate the name. If you pass `name=False` to `da.from_array` the problem goes away. I'll take a look to see if there is an obvious reason that hashing is expanding the array, but even if the memory usage is fixed I would guess the hashing would be very slow since it would need to iterate over the large "virtual" array created by `view_as_windows`. See the documentation of `da.from_array` for an explanation of `name`.

**Comment by bmerry** [comment]
The hash functions operate on a 1D contiguous block of memory. So the only way to avoid the expansion when hashing would be to feed contiguous pieces to it. In general an array might be composed of many small contiguous pieces (at the extreme end, consider `np.ones(10000)[::2]`) which would be terrible for performance. It also looks like `from_array` is designed to accept things that are array-like but not exactly `ndarray`, and trying to do too much clever introspection on them could easily go wrong.

Personally I've never liked the default of trying to hash the argument to `from_array`, and my preferred fix would be to use a time machine to change the default in the past. Without a time machine, I'd suggest that the best that can be done is to improve the documentation, which I can have a look at.

**Comment by TomAugspurger** [comment]
We can also teach `dask.base.tokenize` about strided arrays I think. Is there some combination of a base array + strides that uniquely identifies a strided array? Perhaps something like

```python
In [18]: dask.base.tokenize(strided_array.base.base, strided_array.strides)
Out[18]: 'b49a5b9147024e0564b2532e06e927c8'
```

For reference, 

```python
In [23]: strided_array.base
Out[23]: <numpy.lib.stride_tricks.DummyArray at 0x7f371dcd2490>

In [24]: strided_array.base.base
Out[24]:
array([[       0,        1,        2, ...,     4997,     4998,     4999],
       [    5000,     5001,     5002, ...,     9997,     9998,     9999],
       [   10000,    10001,    10002, ...,    14997,    14998,    14999],
       ...,
       [24985000, 24985001, 24985002, ..., 24989997, 24989998, 24989999],
       [24990000, 24990001, 24990002, ..., 24994997, 24994998, 24994999],
       [24995000, 24995001, 24995002, ..., 24999997, 24999998, 24999999]])
```

So tokenize could check for the presence of `.base`, and if it's a `DummyArray` then tokenize its base + strides (+ anything else needed?)

**Comment by bmerry** [comment]
That might work in theory, but it's a bit scary:
- The numpy docs don't mention DummyArray, so is it guaranteed to continue to exist?
- If it is still exists, is sliding_window_view guaranteed to use it?
- Does it's use guarantee that the current array is a view of the base of the DummyArray? It could be hijacked for other purposes, either by numpy itself or a user.
- It won't detect all cases. For example, `np.broadcast_to` doesn't appear to, and if one slices the result of `sliding_window_view` (even with `[:]`) then `x.base` is not a Dummy (one has to follow the chain of `.base` objects, which again gets fraught if one is using array-likes instead of arrays).
- In general there is no guarantee that the view references the entirety of the base. If not, then irrelevant data will be included in the hash, possibly leading to false mismatches, and possibly making it much *slower* instead of faster.

An alternative would be to directly examine the shape and strides and figure out the memory footprint. That sounds pretty complicated though, and would probably slow things down a lot for small simple arrays.

**Comment by FirefoxMetzger** [comment]
>Is there some combination of a base array + strides that uniquely identifies a strided array?

Would it make sense to do it for general views instead? A view can have a `dtype`, `shape`, `offset`, `strides`, or `order` that is different from the underlying `base` array, while still pointing to the same buffer.

> The numpy docs don't mention DummyArray, so is it guaranteed to continue to exist?

Not exactly, but I'm not a numpy dev, so you may want to take my word with a pinch of salt. When I last dug into this `DummyArray` was something a trick to avoid problems elsewhere. It has been around, but I wouldn't rely on it.

> An alternative would be to directly examine the shape and strides and figure out the memory footprint. That sounds pretty complicated though, and would probably slow things down a lot for small simple arrays.

Yes, this seems like a safer route. It doesn't have to be slower for simple arrays, because they can take a different code path by checking the [`OWNDATA` Flag](https://numpy.org/doc/stable/reference/generated/numpy.ndarray.flags.html).
